File: recency.py
"""Post-RRF recency prior, cross-pack chunk dedup, and version precedence.

Implements the C4 ranking rules frozen by ADR-0004 ("Recency ranking
(chosen here, implemented in C4/#71)" and "Dedup semantics") and issue #71:

    age_months = (now - published_at) / 30.44 days
    multiplier  = 1.0 - (1.0 - floor) * min(1.0, age_months / floor_months)
    adjusted    = rrf_score * multiplier          (applied AFTER fusion)

Shipping default is the LINEAR form (`packs.recency.floor` 0.85,
`packs.recency.floorMonths` 18); the exponential half-life variant
(`packs.recency.halfLifeMonths`, default 9) is reserved and NOT wired (ADR).

Pure functions only: no store, no I/O, no clocks read here (`now` is
injectable so ranking and its tests are deterministic). The same semantics
are implemented for the Node backend in
desktop/main/backend/retrieval/recency.ts; the semver sort key is pinned to
pack_manager._version_key by tests/test_c4_recency_formula.py parity tests.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple

# ADR-0004: the 30.44-day month is part of the pinned formula.
DAYS_PER_MONTH = 30.44
SECONDS_PER_DAY = 86400.0

DEFAULT_FLOOR = 0.85
DEFAULT_FLOOR_MONTHS = 18.0
DEFAULT_HALF_LIFE_MONTHS = 9.0


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


def _as_aware_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


def parse_published_at(value: Any) -> Optional[datetime]:
    """Parse an RFC3339 pack `published_at` (or datetime) to aware UTC.

    Returns None for None/empty/unparseable values so callers render the
    neutral multiplier instead of failing a query over bad metadata.
    """
    if value is None:
        return None
    if isinstance(value, datetime):
        return _as_aware_utc(value)
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        candidate = text.replace("Z", "+00:00") if text.endswith("Z") else text
        try:
            return _as_aware_utc(datetime.fromisoformat(candidate))
        except ValueError:
            return None
    return None


def recency_multiplier(
    published_at: Any = None,
    now: Optional[datetime] = None,
    floor: float = DEFAULT_FLOOR,
    floor_months: float = DEFAULT_FLOOR_MONTHS,
) -> float:
    """Linear recency multiplier r(age_months) per ADR-0004.

    r = 1.0 for age <= 0 (clock skew clamps to fresh); falls linearly from
    1.0 to `floor` across `floor_months`, then holds at `floor`. Missing or
    unparseable `published_at` is neutral (1.0).
    """
    moment = parse_published_at(published_at)
    if moment is None:
        return 1.0
    reference = _as_aware_utc(now) if now is not None else _utc_now()
    age_months = (reference - moment).total_seconds() / (
        DAYS_PER_MONTH * SECONDS_PER_DAY
    )
    if age_months <= 0.0:
        return 1.0
    horizon = float(floor_months) if floor_months else DEFAULT_FLOOR_MONTHS
    fraction = min(1.0, age_months / horizon)
    return 1.0 - (1.0 - float(floor)) * fraction


def semver_sort_key(version: str) -> Tuple[int, int, int, Tuple[int, tuple]]:
    """Semver 2.0.0 section-11 sort key (numeric triple, then pre-release
    ordering per 11.4: numeric identifiers compare numerically and rank
    below alphanumeric ones). Kept in lockstep with
    ``pack_manager._version_key``; parity is asserted by
    tests/test_c4_recency_formula.py so the two cannot drift."""
    core, _, _build = version.partition("+")
    core, _, pre = core.partition("-")
    major_s, minor_s, patch_s = core.split(".")
    major, minor, patch = int(major_s), int(minor_s), int(patch_s)
    if not pre:
        pre_key: Tuple[int, tuple] = (1, ())
    else:
        ids = tuple(
            (0, int(part), "") if part.isdigit() else (1, 0, part)
            for part in pre.split(".")
        )
        pre_key = (0, ids)
    return (major, minor, patch, pre_key)


def _claim_sort_key(claim: Mapping[str, Any]) -> tuple:
    """Precedence key: greater published_at wins, then semver-highest
    version, then lexicographically greatest pack_id (ADR-0004 dedup
    semantics). Published-at ties fall through to the version/id breaks."""
    moment = parse_published_at(claim.get("published_at"))
    epoch = moment.timestamp() if moment is not None else float("-inf")
    version = claim.get("version") or "0.0.0"
    try:
        version_key = semver_sort_key(version)
    except (ValueError, AttributeError):
        version_key = (0, 0, 0, (1, ()))
    pack_id = claim.get("pack_id") or ""
    return (epoch, version_key, pack_id)


def winning_claim(claims: Sequence[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
    """The precedence winner among ACTIVE claims (None if there is none).

    Shared by apply_recency_prior and the VectorStore ranking layer so the
    surviving copy's citation attribution comes from the same precedence
    chain (published_at, semver, pack_id) that picked it.
    """
    active = [c for c in claims if c.get("active")]
    if not active:
        return None
    return max(active, key=_claim_sort_key)


def apply_recency_prior(
    fused_results: Sequence[Tuple[str, float]],
    pack_metadata_by_chunk: Optional[Mapping[str, Sequence[Mapping[str, Any]]]] = None,
    now: Optional[datetime] = None,
    floor: float = DEFAULT_FLOOR,
    floor_months: float = DEFAULT_FLOOR_MONTHS,
) -> List[Tuple[str, float]]:
    """Rank fused RRF results under pack version precedence + recency.

    Args:
        fused_results: ``[(chunk_id, rrf_score), ...]`` in fused order.
        pack_metadata_by_chunk: chunk_id -> claims. Each claim carries
            ``pack_id``, ``version``, ``published_at``, ``active``. Semantics:
              * chunk_id absent from the map -> neutral (kept, multiplier
                1.0, never deduped) -- the unpackaged/legacy rule;
              * present with an EMPTY claim list -> inactive-pack orphan,
                dropped (issue #71 version precedence, defense-in-depth);
              * present with claims -> only ACTIVE claims compete; if none
                is active the chunk is dropped; otherwise exactly one copy
                survives, attributed to the precedence winner (greater
                published_at, then semver-highest version, then
                lexicographically greatest pack_id), scored
                ``rrf_score * recency_multiplier(winner.published_at)``.
    Returns:
        ``[(chunk_id, adjusted_score), ...]`` sorted by adjusted score
        descending; ties keep fused (input) order via a stable sort.
    """
    claims_map = pack_metadata_by_chunk or {}
    # Dedup by chunk identity: exactly one copy survives per chunk_id (the
    # highest adjusted copy), first occurrence wins position ties so the
    # re-ranking is deterministic across repeated identical queries (AC5).
    best: Dict[str, Tuple[float, int]] = {}
    order: List[str] = []
    for position, (chunk_id, score) in enumerate(fused_results):
        claims = claims_map.get(chunk_id)
        if claims is None:
            adjusted = score
        else:
            winner = winning_claim(claims)
            if winner is None:
                # Empty list (orphan) or only inactive claims: excluded from
                # the candidate set entirely (never merely down-weighted).
                continue
            multiplier = recency_multiplier(
                winner.get("published_at"),
                now=now,
                floor=floor,
                floor_months=floor_months,
            )
            adjusted = score * multiplier
        if chunk_id not in best:
            best[chunk_id] = (adjusted, position)
            order.append(chunk_id)
        elif adjusted > best[chunk_id][0]:
            best[chunk_id] = (adjusted, best[chunk_id][1])
    survivors = [(chunk_id, best[chunk_id][0]) for chunk_id in order]
    # Stable sort keeps fused order for equal adjusted scores.
    survivors.sort(key=lambda pair: pair[1], reverse=True)
    return survivors

File: desktop/main/backend/retrieval/recency.ts
// retrieval/recency.ts — post-RRF recency prior, cross-pack chunk dedup, and
// version precedence (C4, issue #71).
//
// Implements the ranking rules frozen by ADR-0004 ("Recency ranking (chosen
// here, implemented in C4/#71)" and "Dedup semantics"):
//
//   ageMonths  = (now - publishedAt) / 30.44 days
//   multiplier = 1.0 - (1.0 - floor) * min(1.0, ageMonths / floorMonths)
//   adjusted   = rrfScore * multiplier        (applied AFTER fusion)
//
// The shipping default is the LINEAR form (packs.recency.floor 0.85,
// packs.recency.floorMonths 18); the exponential half-life variant
// (packs.recency.halfLifeMonths, default 9) is reserved and NOT wired. The
// same semantics live in the Python backend's recency.py — the two MUST NOT
// diverge (ADR-0004 amendment). Semver precedence reuses the PackManager
// comparator (store/pack-manager.ts versionKey) so there is one algorithm.

/** One active pack's claim on a chunk (from the packs/docs join). */
export interface PackClaim {
  packId: string;
  version: string;
  /** Pack manifest publishedAt (RFC3339); null on pre-C4 rows -> neutral. */
  publishedAt: string | null;
  active: boolean;
}

export interface RecencyPriorOptions {
  /** Supply a fixed clock for deterministic ranking/tests. */
  now?: Date;
  /** packs.recency.floor (default 0.85). */
  floor?: number;
  /** packs.recency.floorMonths (default 18). */
  floorMonths?: number;
}

export const DEFAULT_RECENCY_FLOOR = 0.85;
export const DEFAULT_RECENCY_FLOOR_MONTHS = 18;
/** ADR-0004: the 30.44-day month is part of the pinned formula. */
export const DAYS_PER_MONTH = 30.44;

/**
 * Linear recency multiplier r(ageMonths). age <= 0 clamps to fresh (1.0);
 * a missing/unparseable publishedAt is neutral (1.0). Falls linearly from
 * 1.0 to `floor` across `floorMonths`, then holds at the floor.
 */
export function recencyMultiplier(
  publishedAt: string | null | undefined,
  options: RecencyPriorOptions = {},
): number {
  const moment = publishedAtTime(publishedAt);
  if (!Number.isFinite(moment)) return 1;
  const nowMs = (options.now ?? new Date()).getTime();
  const ageMonths = (nowMs - moment) / (DAYS_PER_MONTH * 24 * 60 * 60 * 1000);
  if (ageMonths <= 0) return 1;
  const horizon = options.floorMonths ?? DEFAULT_RECENCY_FLOOR_MONTHS;
  const fraction = Math.min(1, ageMonths / horizon);
  return 1 - (1 - (options.floor ?? DEFAULT_RECENCY_FLOOR)) * fraction;
}

/**
 * Epoch-ms for a pack publishedAt, matching Python parse_published_at:
 * timezone-less ISO datetimes are treated as UTC (Python assigns UTC to
 * naive values) and missing/invalid values map to -Infinity so claim
 * precedence can never promote an unparseable timestamp (Python parity:
 * recency.py _claim_sort_key).
 */
export function publishedAtTime(
  publishedAt: string | null | undefined,
): number {
  if (!publishedAt) return Number.NEGATIVE_INFINITY;
  const text = publishedAt.trim();
  if (!text) return Number.NEGATIVE_INFINITY;
  // Date-only forms ("2024-01-01") are UTC per the JS spec; datetimes
  // without an explicit offset get a Z appended so they parse as UTC too.
  const normalized =
    /[Zz]$|[+-]\d{2}:?\d{2}$/.test(text) || !/T/i.test(text)
      ? text
      : `${text}Z`;
  const ms = Date.parse(normalized);
  return Number.isFinite(ms) ? ms : Number.NEGATIVE_INFINITY;
}

/**
 * Precedence winner among ACTIVE claims: greater publishedAt, then
 * semver-highest version (PackManager comparator — a pre-release sorts below
 * its release), then lexicographically greatest packId (ADR-0004 dedup
 * semantics). Non-active claims never compete.
 */
export function precedenceWinner(claims: readonly PackClaim[]): PackClaim | null {
  const active = claims.filter((claim) => claim.active);
  if (active.length === 0) return null;
  return active.reduce((best, claim) => {
    const bestTime = publishedAtTime(best.publishedAt);
    const claimTime = publishedAtTime(claim.publishedAt);
    if (claimTime !== bestTime) return claimTime > bestTime ? claim : best;
    const versionCmp = safeVersionCompare(claim.version, best.version);
    if (versionCmp !== 0) return versionCmp > 0 ? claim : best;
    return (claim.packId ?? '') > (best.packId ?? '') ? claim : best;
  });
}

/**
 * Rank fused RRF results under pack version precedence + recency.
 *
 * `packMetadataByChunk` maps chunkId -> claims for that chunk:
 *  - chunk absent from the map: NEUTRAL — kept at its fused score, never
 *    deduped (the unpackaged/legacy rule);
 *  - present with an EMPTY claim list: inactive-pack orphan — EXCLUDED from
 *    the candidate set entirely (never merely down-weighted; issue #71 AC2);
 *  - present with claims: only ACTIVE claims compete; no active claim drops
 *    the chunk; otherwise exactly one copy survives, attributed to the
 *    precedence winner, scored `rrfScore * recencyMultiplier(winner)`.
 *
 * Returns `[chunkId, adjustedScore]` sorted by adjusted score DESCENDING,
 * stable on input order for ties, so repeated identical queries rank
 * identically (issue #71 AC5 determinism).
 */
export function applyRecencyPrior(
  fused: ReadonlyArray<readonly [string, number]>,
  packMetadataByChunk: ReadonlyMap<string, readonly PackClaim[]> | null,
  options: RecencyPriorOptions = {},
): Array<[string, number]> {
  // Dedup by chunk identity: exactly one copy survives per chunkId (the
  // highest adjusted copy), first occurrence wins position ties, so the
  // re-ranking is deterministic across repeated identical queries (AC5) —
  // kept in lockstep with the Python apply_recency_prior.
  const best = new Map<string, number>();
  const order: string[] = [];
  for (const [chunkId, score] of fused) {
    const claims = packMetadataByChunk?.get(chunkId);
    let adjusted = score;
    if (claims !== undefined) {
      const winner = precedenceWinner(claims);
      if (winner === null) continue; // orphan / inactive-only: excluded
      adjusted = score * recencyMultiplier(winner.publishedAt, options);
    }
    if (!best.has(chunkId)) {
      best.set(chunkId, adjusted);
      order.push(chunkId);
    } else if (adjusted > (best.get(chunkId) as number)) {
      best.set(chunkId, adjusted);
    }
  }
  const survivors: Array<[string, number]> = order.map((chunkId) => [
    chunkId,
    best.get(chunkId) as number,
  ]);
  survivors.sort((a, b) => b[1] - a[1]);
  return survivors;
}

/**
 * Semver compare that cannot throw into the retrieval path: claims come from
 * the packs table (schema-validated at install), but a malformed row must
 * degrade to "oldest" instead of failing a query.
 */
function safeVersionCompare(a: string, b: string): number {
  try {
    return compareVersionKeys(versionKey(a), versionKey(b));
  } catch {
    return 0;
  }
}

// Imported last for readability; the comparator is the single semver
// authority in the backend (store/pack-manager.ts).
import { compareVersionKeys, versionKey } from '../store/pack-manager.js';
